### PR TITLE
Happy Thoughts API by Linnéa Helén

### PR DESCRIPTION
**Pull request by Polkapetunian**
The deploy wasn't working yesterday but I fixed it today! =D

https://linneas-happy-thoughts.herokuapp.com/thoughts

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -0,0 +1 @@
+MONGO_URL="mongodb+srv://polkaPetun:0sD0oW813Ik0yMXL@cluster0.wjjmc.mongodb.net/happyThoughts?retryWrites=true&w=majority"
```
Comment (katliebig) [inline .env:1]: We're assuming that this is the old connection string from before you listed .env in .gitignore. It's probably still better to remove it.

File: README.md [hunk 2]
```diff
@@ -1,13 +1,25 @@
...
+This is the backend of an app called Happy thoughts. The app allows you to post a thought and like the existing thoughts. I have built the frontend in an earlier project. The purpose of this project was to practise setting up an API that allows users to POST data to a database in MongoDB. It is built using Node.js with Express and MongoDB with Mongoose.
 
 ## The problem
 
-Describe how you approached to problem, and what tools and techniques you used to solve it. How did you plan? What technologies did you use? If you had more time, what would be next?
+I used Mongoose methods from the lectures. The endpoints worked when I tried them locally. Then I had trouble deploying. I got an error trying to look at the Heroku endpoint and I didn't see any collection in my MongoDB Atlas cluster. After much trouble shooting together with a team mate, it turned out I shouldn't have quotes around my connection string in the config vars in Heroku settings. Removing them, it worked! I Also had a little trouble getting the like feature to work when I connected this backend to my frontend. But it turned out there was a missing slash in the endpoint URL.
```
Comment (katliebig) [inline README.md:7]: Good that you solved these issues, but it's not necessary to go into detail here, especially as it was just some typos that caused the problem.

File: README.md [hunk 3]
```diff
@@ -1,13 +1,25 @@
...
+
+https://linneas-happy-thoughts.herokuapp.com//thoughts/:id/likes
+This endpoint doesn't require a JSON body. Given a valid thought id in the URL, the API should find that thought, and update its hearts property to add one heart.
+
+### FULLSTACK PROJECT
+https://happy-thoughts-indeed.netlify.app/ 
+
+Here is the frontend I had made earlier connected to the backend of this project.
+
+
```
Comment (katliebig) [inline README.md:25]: Great that you're listing all available endpoints! You could have a look at the express-list-endpoints package, makes life easier :)

File: server.js [hunk 4]
```diff
@@ -1,28 +1,86 @@
...
+  }
+})
+
+//Adds a like (a heart) to a specifik thought
+app.post('/thoughts/:id/likes', async (req, res) => {
+  const { id } = req.params
+
+  try {
+    const updatedThought = await Thought.findByIdAndUpdate(
+      id,
```
Comment (katliebig) [inline server.js:63]: We were unsure about the formatting here. Guess id can be shortened like that if it's id:id? (but then it's not the mongo-id that's being used?) But it seems to work 😎

**Review by katliebig — APPROVED** [review]
Well done! Good job on the endpoint descriptions. Clean and readable code, and error catching 🤓

File: README.md [hunk 5]
```diff
@@ -1,13 +1,25 @@
...
-Every project should be deployed somewhere. Be sure to include the link to the deployed project so that the viewer can click around and see what it's all about.
+### ENDPOINTS
+
+https://linneas-happy-thoughts.herokuapp.com/thoughts 
+Use with GET to return 20 latest thoughts sorted chronologically by creation time.
+Use with POST to add a thought. This endpoint expects a JSON body with the thought message, like this: { "message": "Express is great!" }.
+
+https://linneas-happy-thoughts.herokuapp.com//thoughts/:id/likes
+This endpoint doesn't require a JSON body. Given a valid thought id in the URL, the API should find that thought, and update its hearts property to add one heart.
+
```
Comment (KarolCislowski) [inline README.md:19]: nice documentation 👍🏻

File: server.js [hunk 6]
```diff
@@ -1,28 +1,86 @@
...
-// Defines the port the app will run on. Defaults to 8080, but can be 
-// overridden when starting the server. For example:
-//
-//   PORT=9000 npm start
+const Thought = mongoose.model('Thought', {
+  message: {
+    type: String,
+    required: true,
+    minlength: 5,
+    maxlength: 140
```
Comment (KarolCislowski) [inline server.js:17]: Nicely configured field 👍🏻

File: server.js [hunk 7]
```diff
@@ -1,28 +1,86 @@
...
 
-// Start defining your routes here
-app.get('/', (req, res) => {
+app.get('/', (_, res) => {
   res.send('Hello world')
 })
 
+//Returns 20 sorted thoughts
+app.get('/thoughts', async (_, res) => {
+  const thoughts = await Thought.find().sort({createdAt: 'desc'}).limit(20).exec()
```
Comment (KarolCislowski) [inline server.js:41]: 👍🏻

File: server.js (hunk 4 above)
Comment (KarolCislowski) [reply server.js:63]: @katliebig - Linnea used a special kind of function `findByIdAndUpdate` which is by default looking for `_id` field.

File: server.js [hunk 8]
```diff
@@ -1,28 +1,86 @@
...
+      id, 
+      { 
+        $inc: {
+          hearts: 1
+        }
+      },
+      {
+        new: true
+      }
+    )
```
Comment (KarolCislowski) [inline server.js:72]: 💯

**Review by KarolCislowski — APPROVED** [review]
Great API 🦄

**Pull request closed without merging**